pkg/alert/alert.go

```diff
@@ -0,0 +1,427 @@
+	Context                     context.Context
+	TestsToExclude              []string
+}
+
+func New(cfg Config) (*Alert, error) {
+	return &Alert{
+		log: cfg.Logger,
+		cfg: cfg,
+		ctx: cfg.Context,
+	}, nil
+}
+
+//FindFailedTests finds distinct, failed tests, that has not yet been posted on slack in recent n days
+func (alert *Alert) FindFailedTests() map[string]TestDetails {
+	testAggregationsRaw := alert.retrieveTestAggregations()
+	contextToTestDetailMap := extractTestDetailItems(testAggregationsRaw)
+	failedTests := alert.removeSuccessfulTests(contextToTestDetailMap)
+	alert.removeExcludedTests(failedTests)
+	alert.removeAlreadyFiledAlerts(failedTests)
+	return failedTests
+}
+
+func (alert *Alert) retrieveTestAggregations() testContextAggregation {
+	payloadFormated := fmt.Sprintf(`{
+		"size": 0,
+		"query": {
+			"bool": {
+				"must": [
+					{ "match": { "type": "teststep" } },
+					{ "range": { "startTime": { "gte": "now-%dd" } } }
+				],
+				"should": [
+					{ "term": { "phase.keyword": "Failed" } },
+					{ "term": { "phase.keyword": "Succeeded" } },
+					{ "term": { "phase.keyword": "Skipped" } }
+				],
+				"minimum_should_match": 1
+			}
+		},
+		"aggs": {
+			"test_context": {
+				"terms": {
+					"script": {
+						"source": "def landscape = ''; def k8s_version = ''; def name = ''; def provider = 'none'; def os = ''; if (doc['tm.cloudprovider.keyword'].size() != 0) { provider = doc['tm.cloudprovider.keyword'].value;  } if (doc['tm.landscape.keyword'].size() != 0) { landscape = doc['tm.landscape.keyword'].value; } if (doc['tm.operating_system.keyword'].size() != 0) { os = doc['tm.operating_system.keyword'].value; } if (doc['tm.k8s_version.keyword'].size() != 0) { k8s_version = '_v' + doc['tm.k8s_version.keyword'].value; } if (doc['name.keyword'].size() != 0) { name = doc['name.keyword'].value; } name + '_' + landscape + '_' + provider + k8s_version + '_' + os;",
+						"lang": "painless"
+					},
+					"min_doc_count": %d,
+					"size": 10000
+				},
+				"aggs": {
+					"success_rate": { "avg": { "field": "pre.phaseNum" } },
+					"success_trend": {
+						"top_hits": {
+							"sort": [ { "startTime": { "order": "desc" } } ],
+							"_source": { "includes": [ "pre.phaseNum" ] },
+							"size": %d
+						}
+					},
+					"details": {
+						"top_hits": {
+							"sort": [ { "startTime": { "order": "desc" } } ],
+							"_source": {
+								"includes": [
+									"name",
+									"startTime",
+									"pre.clusterDomain",
+									"tm.landscape",
+									"tm.tr.id",
+									"tm.cloudprovider",
+									"tm.k8s_version",
+									"tm.operating_system"
+								]
+							},
+							"size": 1
+						}
+					}
+				}
+			}
+		}
+	}`, alert.cfg.EvalTimeDays, alert.cfg.ContinuousFailureThreshold, alert.cfg.ContinuousFailureThreshold)
+
+	var testContextAggregation testContextAggregation
+	alert.elasticRequest(`/testmachinery-*/_search`, "GET", payloadFormated, &testContextAggregation)
```

Comment:
use single line strings `"`
```suggestion
	alert.elasticRequest("/testmachinery-*/_search", "GET", payloadFormated, &testContextAggregation)
```

Reply:
```suggestion
	alert.elasticRequest(`/testmachinery-*/_search`, http.MethodGet, payloadFormated, &testContextAggregation)
```
use constant https://golang.org/pkg/net/http/#pkg-constants